### PR TITLE
[Federation][(Un)join-00] Implement federation/cluster resource generator.

### PR DESCRIPTION
**Pull request by madhusudancs**
Design Doc: PR #34484

cc @kubernetes/sig-cluster-federation @nikhiljindal

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35153)

### CODE REVIEW

File: pkg/kubectl/cluster.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+	"k8s.io/kubernetes/pkg/runtime"
+)
+
+// ClusterGeneratorV1Beta1 supports stable generation of a
+// federation/cluster resource.
+type ClusterGeneratorV1Beta1 struct {
+	// Name of the cluster context (required)
+	Name string
+	// ServerAddress is the APIServer address of the Kubernetes cluster
+	// that is being registered (optional)
```
Comment (nikhiljindal) [inline pkg/kubectl/cluster.go]: Why optional? What happens when this is not specified? Is there a default value?

File: pkg/kubectl/cluster.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+	ServerAddress string
+	// SecretName is the name of the secret that stores the credentials
+	// for the Kubernetes cluster that is being registered (optional)
```
Comment (nikhiljindal) [inline pkg/kubectl/cluster.go]: Same question for what happens when not specified

File: pkg/kubectl/cluster.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+		return nil, err
+	}
+	cluster := &federationapi.Cluster{
+		ObjectMeta: v1.ObjectMeta{
+			Name: s.Name,
+		},
+		Spec: federationapi.ClusterSpec{
+			ServerAddressByClientCIDRs: []federationapi.ServerAddressByClientCIDR{
+				{
+					ServerAddress: s.ServerAddress,
```
Comment (nikhiljindal) [inline pkg/kubectl/cluster.go]: I dont think ClientCIDR is optional

File: pkg/kubectl/cluster_test.go [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+	tests := []struct {
+		params    map[string]interface{}
+		expected  *federationapi.Cluster
+		expectErr bool
+	}{
+		{
+			params: map[string]interface{}{
+				"name": "foo",
+			},
+			expected: &federationapi.Cluster{
```
Comment (nikhiljindal) [inline pkg/kubectl/cluster_test.go]: I am not sure if this is a valid cluster spec, that the apiserver will accept

File: pkg/kubectl/cluster.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+	clustergen.SecretName = params["secret"]
+	return clustergen.StructuredGenerate()
+}
+
+// ParamNames returns the set of supported input parameters when using
+// the parameter injection generator pattern.
+func (s ClusterGeneratorV1Beta1) ParamNames() []GeneratorParam {
+	return []GeneratorParam{
+		{"name", true},
+		{"serverAddress", false},
```
Comment (nikhiljindal) [inline pkg/kubectl/cluster.go]: Is this a command line flag? Should it be `server-address` in that case?

File: pkg/kubectl/cluster.go (hunk 3 above)
Comment (madhusudancs) [inline pkg/kubectl/cluster.go]: Fixed.

File: pkg/kubectl/cluster_test.go (hunk 4 above)
Comment (madhusudancs) [inline pkg/kubectl/cluster_test.go]: Fixed.

**Review by madhusudancs — COMMENTED** [review]
@nikhiljindal addressed the comments. PTAL.

File: pkg/kubectl/cluster.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+// validate validates required fields are set to support structured
+// generation.
+func (s ClusterGeneratorV1Beta1) validate() error {
+	if len(s.Name) == 0 {
+		return fmt.Errorf("name must be specified")
+	}
+	if len(s.ClientCIDR) == 0 {
+		return fmt.Errorf("client CIDR must be specified")
```
Comment (nikhiljindal) [inline pkg/kubectl/cluster.go]: We can use "0.0.0.0/0" as the default CIDR (matches everything)

File: pkg/kubectl/cluster.go [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+	}
+	if len(s.ServerAddress) == 0 {
+		return fmt.Errorf("server address must be specified")
+	}
+	if len(s.SecretName) == 0 {
+		return fmt.Errorf("secret name must be specified")
```
Comment (nikhiljindal) [inline pkg/kubectl/cluster.go]: We can use cluster name as default secret name

File: pkg/kubectl/cluster.go [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+	return clustergen.StructuredGenerate()
+}
+
+// ParamNames returns the set of supported input parameters when using
+// the parameter injection generator pattern.
+func (s ClusterGeneratorV1Beta1) ParamNames() []GeneratorParam {
+	return []GeneratorParam{
+		{"name", true},
+		{"client-cidr", false},
+		{"server-address", false},
```
Comment (nikhiljindal) [inline pkg/kubectl/cluster.go]: Shouldnt the second parameter be true here?

**Comment by nikhiljindal** [comment]
I think cluster name and server-address should only be required. secretName and clientCIDR can be optional. Looks good otherwise

File: pkg/kubectl/cluster.go (hunk 8 above)
Comment (madhusudancs) [inline pkg/kubectl/cluster.go]: Done.

File: pkg/kubectl/cluster.go (hunk 6 above)
Comment (madhusudancs) [inline pkg/kubectl/cluster.go]: My original plan was to do this defaulting in the caller. It doesn't hurt to default at both the places. So done.

File: pkg/kubectl/cluster.go (hunk 7 above)
Comment (madhusudancs) [inline pkg/kubectl/cluster.go]: Same as above. Done.

**Review by madhusudancs — COMMENTED** [review]
@nikhiljindal addressed the comments. PTAL.

**Comment by nikhiljindal** [comment]
Thx for addressing the comments.
LGTM.
Please squash commits.

**Comment by madhusudancs** [comment]
@nikhiljindal squashed the commits and rebased. Thank you very much for a thorough review.

**Comment by madhusudancs** [comment]
@k8s-bot  node e2e test this

**Comment by madhusudancs** [comment]
@k8s-bot node e2e test this

**Comment by madhusudancs** [comment]
@k8s-bot unit test this

**Comment by madhusudancs** [comment]
@k8s-bot unit test this

**Comment by madhusudancs** [comment]
@k8s-bot unit test this

**Comment by nikhiljindal** [comment]
@k8s-bot unit test this

**Comment by madhusudancs** [comment]
@k8s-oncall pinging k8s-bot doesn't seem to work. Any idea what's going on?

**Comment by saad-ali** [comment]
@k8s-bot unit test this

**Comment by saad-ali** [comment]
@kubernetes/test-infra-admins Any ideas why k8s-bot is not responding?

**Comment by saad-ali** [comment]
@k8s-bot unit test this, Issue #35898

**Comment by saad-ali** [comment]
Manually merging to unblock--failing unit test is due to #35898. All other tests are green.

**Comment by madhusudancs** [comment]
@saad-ali thanks!

**Comment by ixdy** [comment]
it looks like the tests were run several times (per https://k8s-gubernator.appspot.com/pr/35153), but I don't know why it didn't update the commit status here. @spxtr any ideas?

**Pull request merged**